py/desispec/exposure_qa.py

```diff
@@ -0,0 +1,377 @@
+
+    petal_tsnr2=np.zeros(10)
+    worst_rdnoise = 0
+
+    fiberqa_table["EFFTIME_SPEC"]=np.zeros(fiberqa_table["TARGETID"].size)
+
+    petalqa_table = Table()
+    npetal=10
+    petalqa_table["PETAL_LOC"]=np.arange(npetal)
+    petalqa_table["WORSTREADNOISE"]=np.zeros(npetal,dtype=float)
+    petalqa_table["NGOODPOS"]=np.zeros(npetal,dtype=int)
+    petalqa_table["NGOODFIB"]=np.zeros(npetal,dtype=int)
+    petalqa_table["NSTDSTAR"]=np.zeros(npetal,dtype=int)
+    petalqa_table["STARRMS"]=np.zeros(npetal,dtype=float)
+    petalqa_table["TSNR2FRA"]=np.zeros(npetal,dtype=float)
+    petalqa_table["EFFTIME_SPEC"]=np.zeros(npetal,dtype=float)
+    petalqa_table["NCFRAME"]=np.zeros(npetal,dtype=int)
+    petalqa_table["BSKYTHRURMS"]=np.zeros(npetal,dtype=float)
+    petalqa_table["BSKYCHI2PDF"]=np.zeros(npetal,dtype=float)
+    petalqa_table["RSKYTHRURMS"]=np.zeros(npetal,dtype=float)
+    petalqa_table["RSKYCHI2PDF"]=np.zeros(npetal,dtype=float)
+    petalqa_table["ZSKYTHRURMS"]=np.zeros(npetal,dtype=float)
+    petalqa_table["ZSKYCHI2PDF"]=np.zeros(npetal,dtype=float)
+    petalqa_table["BTHRUFRAC"]=np.zeros(npetal,dtype=float)
+    petalqa_table["RTHRUFRAC"]=np.zeros(npetal,dtype=float)
+    petalqa_table["ZTHRUFRAC"]=np.zeros(npetal,dtype=float)
+
+    # need to add things
+
+    frame_header = None
+    fibermap_header = None
+
+    # EFFTIME
+    goaltype="dark"
+    if fibermap_header is not None and "GOALTYPE" in fibermap_header :
+        goaltype=fibermap_header["GOALTYPE"]
+    else :
+        log.warning("no GOALTYPE info, assume 'dark'")
+    param_name="tsnr2_for_efftime_{}".format(goaltype.lower())
+    if param_name in qa_params :
+        tsnr2_for_efftime_key = qa_params[param_name]
+    else :
+        tsnr2_for_efftime_key = "TSNR2_ELG"
+        log.warning("no parameter '{}', use '{}'".format(param_name,tsnr2_for_efftime_key))
+
+    for petal in petal_locs :
+        worst_rdnoise_per_petal = 0
+
+        spectro=petal # same number
+        log.info("spectro {}".format(spectro))
+        entries = np.where(fiberqa_table['PETAL_LOC'] == petal)[0]
+
+        # checking readnoise level
+        ####################################################################
+        bad_rdnoise_mask = fibermask.mask('BADREADNOISE')
+        max_rdnoise      = qa_params["max_readnoise"]
+        for band in ["b","r","z"] :
+            camera=f"{band}{spectro}"
+            cframe_filename=findfile('cframe',night,expid,camera,specprod_dir=specprod_dir)
+            if not os.path.isfile(cframe_filename) :
+                continue
+            petalqa_table["NCFRAME"][petal]+=1
+            head=fitsio.read_header(cframe_filename)
+            if frame_header is None :
+                frame_header = head
+            if fibermap_header is None :
+                try :
+                    fibermap_header = fitsio.read_header(cframe_filename,"FIBERMAP")
+                except OSError as e :
+                    log.error(e)
```

Comment:
This looks like a debugging leftover.  If a cframe doesn't have a FIBERMAP HDU, then something is really wrong with the processing.  If you are going to use the original fibermap file (instead of the cframe FIBERMAP HDU), then it seems you should also use that header.  Otherwise if you want to use the cframe FIBERMAP instead, you should probably use that for everything and require it to exist.

Reply:
I now use directly fibermap.meta.